pkg/adaptor/hypervisor/azure/server.go

```diff
@@ -0,0 +1,106 @@
+//go:build azure
+
+package azure
+
+import (
+	"context"
+	"log"
+	"net"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/hypervisor"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/podnetwork"
+	"github.com/containerd/ttrpc"
+
+	pb "github.com/kata-containers/kata-containers/src/runtime/protocols/hypervisor"
+)
+
+var logger = log.New(log.Writer(), "[helper/hypervisor] ", log.LstdFlags|log.Lmsgprefix)
+
+type server struct {
+	socketPath string
+
+	ttRpc   *ttrpc.Server
+	service pb.HypervisorService
+
+	workerNode podnetwork.WorkerNode
+
+	readyCh  chan struct{}
+	stopCh   chan struct{}
+	stopOnce sync.Once
+}
+
+func NewServer(cfg hypervisor.Config, cloudCfg Config, workerNode podnetwork.WorkerNode, daemonPort string) hypervisor.Server {
+
+	logger.Printf("hypervisor config %v", cfg)
+	logger.Printf("cloud config %v", cloudCfg)
+	azureClient, err := NewAzureClient(cloudCfg)
+	if err != nil {
+		return nil
+	}
+
+	return &server{
+		socketPath: cfg.SocketPath,
+		service:    newService(azureClient, &cloudCfg, &cfg, workerNode, cfg.PodsDir, daemonPort),
+		workerNode: workerNode,
+		readyCh:    make(chan struct{}),
+		stopCh:     make(chan struct{}),
+	}
+}
+
+func (s *server) Start(ctx context.Context) (err error) {
+
+	ttRpc, err := ttrpc.NewServer()
+	if err != nil {
+		return err
+	}
+	s.ttRpc = ttRpc
```

Comment:
I am not familiar with ttRpc server.
As I see it, you need to call ttRpc.Shutdown() on it after allocation, but the "defer func()" that calls s.ttRpc.Shutdown() is set later on. So on the following return error paths (MkdirAll(), RemovAll() and net.Listen()), the ttRpc server is not deallocated.
Is it because Serve() was not called on it yet?

Reply:
@littlejawa that's my understanding as well. I think it'll be good to have a separate issue to go through the ttRpc server cleanup for all the providers.

Reply:
Did a similar exercise w.r.to using the right `context` - https://github.com/confidential-containers/cloud-api-adaptor/issues/131

Reply:
https://github.com/confidential-containers/cloud-api-adaptor/issues/172